cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/io/cdap/cdap/etl/batch/StructuredRecordComparator.java

```diff
@@ -0,0 +1,161 @@
+/*
+ * Copyright © 2019 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.etl.batch;
+
+import io.cdap.cdap.api.common.Bytes;
+import io.cdap.cdap.api.data.format.StructuredRecord;
+import io.cdap.cdap.api.data.schema.Schema;
+
+import java.nio.ByteBuffer;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Comparator;
+import java.util.Iterator;
+
+/**
+ * Compares StructuredRecords. Records cannot contain unions that do not represent a nullable type and they cannot
+ * contain maps. Arrays cannot contain non-comparable types.
+ *
+ * Records are first compared by their schemas. If their schemas are the same, they are compared field by field.
+ *
+ * Record fields are compared first by their schema fields. If the schema fields are the same, they are compared by
+ * value.
+ *
+ * Numbers, boolean, and strings are compared with their normal java compare methods. A null value is less than
+ * a non-null value.
+ *
+ * Arrays are compared by their values in order.
+ *
+ * Bytes are compared lexicographically.
+ */
+public class StructuredRecordComparator implements Comparator<StructuredRecord> {
+  private static final Comparator<Schema> SCHEMA_COMPARATOR = new SchemaComparator();
+
+  @Override
+  public int compare(StructuredRecord r1, StructuredRecord r2) {
+    if (r1 == null && r2 != null) {
+      return -1;
+    } else if (r1 != null && r2 == null) {
+      return 1;
+    } else if (r1 == null) {
+      return 0;
+    }
+
+    int comp = SCHEMA_COMPARATOR.compare(r1.getSchema(), r2.getSchema());
```

Comment:
is the schema comparator necessary? If they have different schema, then two records are different, right? So perhaps the schema comparison is unnecessary?

Reply:
oh the schema is used to determine how to compare fields. Probably then that is not a good suggestion.